### PR TITLE
PR: Remove comments from the function definition lines before generating docstrings.

### LINKED ISSUE #21527: Generate docstring not working if there are comments on the function definition lines
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
On master branch generating docstrings can fail or result in formatting issues of the docstring if function definition lines are commented. This can be fixed with some simple regex, and I will submit a PR for that.

### What steps reproduce the problem?
Two examples
```python
# Not able to generate docstring if the end of function definition line is commented
def test1(v: str = "#"):  # comment, with '#' and "#"

    ...
    
    
    
# Formatting issues with comments, and commas in comments add extra arguments
def test2(v1: str = "#", # comment, with '#' and "#"
          v2: str = '#') -> str:
    """
    

    Parameters
    ----------
    v1 : str, optional
        DESCRIPTION. The default is "#".
    # comment : TYPE
        DESCRIPTION.
    with '#' and "#"          v2 : str, optional
        DESCRIPTION. The default is '#'.

    Returns
    -------
    str
        DESCRIPTION.

    """

    ...
    
```

### Paste Traceback/Error Below (if applicable)

```python-traceback

```

## Versions

* Spyder version: 6.0.0.dev0 11b9e503e (conda)
* Python version: 3.10.13 64-bit
* Qt version: 5.15.8
* PyQt5 version: 5.15.9
* Operating System: Windows-10-10.0.19045-SP0

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0              :  1.4.1 (OK)
chardet >=2.0.0                   :  5.2.0 (OK)
cloudpickle >=0.5.0               :  3.0.0 (OK)
cookiecutter >=1.6.0              :  2.4.0 (OK)
diff_match_patch >=20181111       :  20230430 (OK)
intervaltree >=3.0.2              :  3.1.0 (OK)
IPython >=8.13.0,<9.0.0,!=8.17.1  :  8.17.2 (OK)
jedi >=0.17.2,<0.20.0             :  0.19.1 (OK)
jellyfish >=0.7                   :  1.0.1 (OK)
jsonschema >=3.2.0                :  4.19.2 (OK)
keyring >=17.0.0                  :  24.2.0 (OK)
nbconvert >=4.0                   :  7.11.0 (OK)
numpydoc >=0.6.0                  :  1.5.0 (OK)
paramiko >=2.4.0                  :  3.3.1 (OK)
parso >=0.7.0,<0.9.0              :  0.8.3 (OK)
pexpect >=4.4.0                   :  4.8.0 (OK)
pickleshare >=0.4                 :  0.7.5 (OK)
psutil >=5.3                      :  5.9.5 (OK)
pygments >=2.0                    :  2.16.1 (OK)
pylint >=2.5.0,<3.1               :  3.0.2 (OK)
pylint_venv >=3.0.2               :  3.0.3 (OK)
pyls_spyder >=0.4.0               :  0.4.0 (OK)
pylsp >=1.9.0,<1.10.0             :  1.9.0 (OK)
pylsp_black >=1.2.0,<3.0.0        :  1.3.0 (OK)
pyuca >=1.2                       :  1.2 (OK)
qdarkstyle >=3.2.0,<3.3.0         :  3.2 (OK)
qstylizer >=0.2.2                 :  0.2.2 (OK)
qtawesome >=1.2.1                 :  1.2.3 (OK)
qtconsole >=5.5.0,<5.6.0          :  5.6.0.dev0 (OK)
qtpy >=2.1.0                      :  2.4.1 (OK)
rtree >=0.9.7                     :  1.1.0 (OK)
setuptools >=49.6.0               :  68.2.2 (OK)
sphinx >=0.6.6                    :  7.2.6 (OK)
spyder_kernels >=3.0.0b2,<3.0.0b3 :  3.0.0.dev0 (OK)
superqt >=0.6.1,<1.0.0            :  0.6.1 (OK)
textdistance >=4.2.0              :  4.5.0 (OK)
three_merge >=0.1.1               :  0.1.1 (OK)
watchdog >=0.10.3                 :  3.0.0 (OK)
zmq >=22.1.0                      :  25.1.1 (OK)

# Optional:
cython >=0.21                     :  3.0.5 (OK)
matplotlib >=3.0.0                :  3.8.1 (OK)
numpy >=1.7                       :  1.26.0 (OK)
pandas >=1.1.1                    :  2.1.2 (OK)
scipy >=0.17.0                    :  1.11.3 (OK)
sympy >=0.7.3                     :  1.12 (OK)

```

### PR DESCRIPTION
**Pull request by rhkarls**
## Description of Changes

Remove comments from the function definition lines before generating docstrings. Comments are removed, but # symbols inside quotes are kept as these might be default function arguments.

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Behaviour before:
issue_before_fix

Behaviour with this PR:
issue_after_fix

### Issue(s) Resolved

Fixes #21527

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @rhkarls

### CODE REVIEW

**Comment by rhkarls** [comment]
Realized I missed creating a branch locally...sorry about that. Can that be fixed after pushing, or better create a new pr?

File: spyder/plugins/editor/extensions/docstring.py [hunk 1]
```diff
@@ -120,6 +120,12 @@ def is_in_scope_backward(text):
         text.replace(r"\"", "").replace(r"\'", "")[::-1])
 
 
+def remove_comments(text):
+    """Remove code comments from text while ignoring #-symbols inside quotes,
+    which can be part of function arguments."""
```
Comment (ccordoba12) [inline spyder/plugins/editor/extensions/docstring.py:125]:
```suggestion
    """
    Remove code comments from text while ignoring hash symbols (#) inside
    quotes, which can be part of function arguments.
    """
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks for your work on this @rhkarls! And for the test too.

All seems pretty good, except for the small suggestion I left below.

**Comment by ccordoba12** [comment]
> Realized I missed creating a branch locally...sorry about that. Can that be fixed after pushing, or better create a new pr?

Oops, I missed this. I think you need to create a new PR because there are some fixes in our master branch that you need to get here to make our tests pass.

And merging Spyder's master with your master could cause ugly problems in your fork.

**Comment by rhkarls** [comment]
> > Realized I missed creating a branch locally...sorry about that. Can that be fixed after pushing, or better create a new pr?
> 
> Oops, I missed this. I think you need to create a new PR because there are some fixes in our master branch that you need to get here to make our tests pass.
> 
> And merging Spyder's master with your master could cause ugly problems in your fork.

Thanks for the feedback! I will fix the new PR when I have a free moment, and include your comment

**Pull request closed without merging**